File: C/combinatoric/permutation.c
#include <stdio.h>
#include <string.h>

// Function to swap characters at position i and j in a string
void swap(char* str, int i, int j) {
    char temp = str[i];
    str[i] = str[j];
    str[j] = temp;
}

// Recursive function to generate all permutations of a string
void permute(char* str, int start, int end) {
    if (start == end) {
        printf("%s\n", str); // Print the current permutation
    } else {
        for (int i = start; i <= end; i++) {
            // Swap characters at positions start and i
			swap(str, start, i);
			// Recursively generate permutations for the remaining characters
            permute(str, start + 1, end);
			// Backtrack by restoring the original order 
            swap(str, start, i);      
        }
    }
}

int main() {
    char input[100];
    printf("Enter a string: ");
    scanf("%s", input);

    int length = strlen(input);

    permute(input, 0, length - 1);

    return 0;
}

File: Python/permutation.py
def permute(nums):
    def backtrack(start):
        if start == len(nums):
            # When we have generated a full permutation, add it to the result.
            result.append(nums[:])
            return

        for i in range(start, len(nums)):
            # Swap the current element with the element at index 'start'.
            nums[start], nums[i] = nums[i], nums[start]

            # Recursively generate permutations for the remaining elements.
            backtrack(start + 1)

            # Restore the original order (backtrack).
            nums[start], nums[i] = nums[i], nums[start]

    result = []
    backtrack(0)
    return result

# Example usage:
nums = [1, 2, 3]
print(permute(nums))  # Should return all possible permutations of [1, 2, 3]
